File: OOP3Behav14.cpp
#include "pt4.h"
using namespace std;

class Request
{
public:
    virtual string ToStr() = 0;
};

// Implement the RequestA and RequestB descendant classes
class RequestA : public Request
{
public:
    RequestA(int param) : param(param) {}
    int GetParam() { return param; }
    string ToStr() override { return string("A:") + to_string(param); }

private:
    int param;
};

class RequestB : public Request
{
public:
    RequestB(string param) : param(param) {}
    string GetParam() { return param; }
    string ToStr() override { return string("B:") + param; }

private:
    string param;
};

class Handler
{
    Handler* successor;
public:
    Handler(Handler* successor) : successor(successor) {}
    virtual void HandleRequest(Request* req);
    ~Handler();
};

void Handler::HandleRequest(Request* req)
{
    // Implement the method
    if (successor)
        successor->HandleRequest(req);
    else
        pt << string("Request ") + req->ToStr() + " not processed";
}
Handler::~Handler()
{
    // Implement the destructor
    if (successor)
        successor->~Handler();
}

// Implement the HandlerA and HandlerB descendant classes
class HandlerA : public Handler
{
public:
    HandlerA(Handler *successor, int id, int param1, int param2) : Handler(successor), id(id), param1(param1), param2(param2) {}
    void HandleRequest(Request *req) override;

private:
    int id;
    int param1, param2;
};

void HandlerA::HandleRequest(Request *req)
{
    string flag = req->ToStr();
    if (flag[0] == 'A')
    {
        int param = stoi(flag.substr(2));
        if (param >= param1 && param <= param2)
            pt << string("Request " + flag + " processed by handler " + to_string(id));
        else
            Handler::HandleRequest(req);
    }
    else
        Handler::HandleRequest(req);
}

class HandlerB : public Handler
{
public:
    HandlerB(Handler *successor, int id, string param1, string param2) : Handler(successor), id(id), param1(param1), param2(param2) {}
    void HandleRequest(Request *req) override;

private:
    int id;
    string param1, param2;
};

void HandlerB::HandleRequest(Request *req)
{
    string flag = req->ToStr();
    if (flag[0] == 'B')
    {
        string param = flag.substr(2);
        if (param >= param1 && param <= param2)
            pt << string("Request " + flag + " processed by handler " + to_string(id));
        else
            Handler::HandleRequest(req);
    }
    else
        Handler::HandleRequest(req);
}

class Client
{
    Handler* h;
public:
    Client(Handler* h) : h(h) {}
    void SendRequest(Request* req);
    ~Client();
};

void Client::SendRequest(Request* req)
{
    h->HandleRequest(req);
}
Client::~Client()
{
    // Implement the destructor
    if (h){
    	h->~Handler();
	}
}

void Solve()
{
    Task("OOP3Behav14");
    int n;
    pt >> n;
    Handler *h = new Handler(nullptr);
    for (int i = 0; i < n; i++)
    {
        char flag;
        pt >> flag;
        if (flag == 'A')
        {
            int num1, num2;
            pt >> num1 >> num2;
            h = new HandlerA(h, i, num1, num2);
        }
        else
        {
            string s1, s2;
            pt >> s1 >> s2;
            h = new HandlerB(h, i, s1, s2);
        }
    }
    Client cl(h);
    int k;
    pt >> k;
    for (int i = 0; i < k; i++)
    {
        char c;
        pt >> c;
        if (c == 'A')
        {
            int num;
            pt >> num;
            cl.SendRequest(new RequestA(num));
        }
        else
        {
            string s;
            pt >> s;
            cl.SendRequest(new RequestB(s));
        }
    }
}

File: OOP3Behav14.cs
// File: "OOP3Behav14"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public abstract class Request
        {
            public abstract string ToStr();
            public abstract string GetParam();
        }

        // Implement the RequestA and RequestB descendant classes
        public class RequestA : Request
        {
            public int p;
            public RequestA(int p)
            {
                this.p = p;
            }
            public override string GetParam()
            {
                return p.ToString();
            }
            public override string ToStr()
            {
                return "A:" + p;
            }
        }

        public class RequestB : Request
        {
            public string p;
            public RequestB(string p)
            {
                this.p = p;
            }

            public override string GetParam()
            {
                return p;
            }

            public override string ToStr()
            {
                return "B:" + p;
            }
        }
        public class Handler
        {
            Handler successor;
            public Handler(Handler successor)
            {
                this.successor = successor;
            }
            public virtual void HandleRequest(Request req)
            {
                // Implement the method
                if (successor != null)
                    successor.HandleRequest(req);
                else
                    Put("Request " + req.ToStr() + " not processed");
            }
        }

        // Implement the HandlerA and HandlerB descendant classes
        public class HandlerA : Handler
        {
            int idnt, p1, p2;

            public HandlerA(Handler successor, int id, int p1, int p2) : base(successor)
            {
                this.idnt = id;
                this.p1 = p1;
                this.p2 = p2;
            }
            public override void HandleRequest(Request r)
            {
                if (r.ToStr()[0] == 'A')
                    if (Convert.ToInt32(r.GetParam()) >= p1 && Convert.ToInt32(r.GetParam()) <= p2)
                    {
                        Put("Request " + r.ToStr() + " processed by handler " + idnt);
                        return;
                    }
                base.HandleRequest(r);
            }
        }

        public class HandlerB : Handler
        {
            int id;
            string p1, p2;

            public HandlerB(Handler successor, int id, string p1, string p2) : base(successor)
            {
                this.id = id;
                this.p1 = p1;
                this.p2 = p2;
            }
            public override void HandleRequest(Request req)
            {
                if (req.ToStr()[0] == 'B')
                {
                    if (string.CompareOrdinal(req.GetParam(), p1) >= 0 && string.CompareOrdinal(req.GetParam(), p2) <= 0)
                    {
                        Put("Request " + req.ToStr() + " processed by handler " + id);
                        return;
                    }
                }
                base.HandleRequest(req);
            }
        }
        public class Client
        {
            Handler h;
            public Client(Handler h)
            {
                this.h = h;
            }
            public void SendRequest(Request req)
            {
                h.HandleRequest(req);
            }
        }

        public static void Solve()
        {
            Task("OOP3Behav14");
            int n = GetInt();
            Handler h = new Handler(null);
            for (int i = 0; i < n; i++)
            {
                char c = GetChar();
                if (c == 'A') { h = new HandlerA(h, i, GetInt(), GetInt()); }
                else if (c == 'B') { h = new HandlerB(h, i, GetString(), GetString()); }
            }
            Client cli = new Client(h);
            int k = GetInt();
            for (int i = 0; i < k; i++)
            {
                char c = GetChar();
                if (c == 'A') { cli.SendRequest(new RequestA(GetInt())); }
                else if (c == 'B') { cli.SendRequest(new RequestB(GetString())); }
            }
        }
    }
}
